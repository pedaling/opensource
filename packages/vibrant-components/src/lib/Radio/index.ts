export { Radio } from './Radio';
export { useRadio } from './useRadio';
