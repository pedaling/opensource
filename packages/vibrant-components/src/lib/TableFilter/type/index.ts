export type StringFilterOperator = 'contains' | 'empty' | 'equals' | 'notContains' | 'notEmpty' | 'notEquals';

export type StringFilter = {
  value: string;
  operator: StringFilterOperator[];
};

export type Filter = StringFilter & { dataKey: string };
