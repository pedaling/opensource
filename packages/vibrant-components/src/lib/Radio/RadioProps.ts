import type { Ref } from 'react';
import type { ReactElementChild, ResponsiveValue } from '@vibrant-ui/core';
import { propVariant, withVariation } from '@vibrant-ui/core';

export type RadioProps = {
  ref?: Ref<any>;
  testId?: string;
  children?: ReactElementChild;
  value: string;
  checked?: boolean;
  size?: ResponsiveValue<'md' | 'sm'>;
  direction?: ResponsiveValue<'horizontal' | 'vertical'>;
  disabled?: boolean;
  label?: string;
  description?: string;
  tabIndex?: number;
};

export const withRadioVariation = withVariation<RadioProps>('Radio')(
  propVariant({
    props: [
      {
        name: 'direction',
        responsive: true,
        default: 'vertical',
        keep: true,
      },
    ],
    variants: {
      vertical: {
        flexDirection: 'column',
        width: '100%',
      },
      horizontal: {
        flexDirection: 'row',
        width: 'auto',
      },
    } as const,
  })
);

export const withRadioIconVariation = withVariation<Pick<RadioProps, 'checked' | 'disabled' | 'size'>>('RadioIcon')(
  propVariant({
    props: [
      {
        name: 'size',
        responsive: true,
        default: 'md' as const,
      },
    ],
    variants: {
      md: {
        size: 24,
      },
      sm: {
        size: 20,
      },
    },
  })
);

export const withRadioLabelVariation = withVariation<Pick<RadioProps, 'disabled' | 'label' | 'size'>>('RadioLabel')(
  propVariant({
    props: [
      {
        name: 'size',
        responsive: true,
        default: 'md' as const,
      },
    ],
    variants: {
      md: {
        bodyLevel: 2,
        pl: 8,
        pt: 3,
      },
      sm: {
        bodyLevel: 4,
        pl: 8,
        pt: 2,
      },
    } as const,
  })
);

export const withRadioDescriptionVariation = withVariation<
  Pick<RadioProps, 'description' | 'direction' | 'disabled' | 'size'>
>('RadioDescription')(
  propVariant({
    props: [
      {
        name: 'size',
        responsive: true,
        default: 'md' as const,
      },
      {
        name: 'direction',
        responsive: true,
        default: 'vertical' as const,
      },
    ],
    variants: ({ size, direction }) => {
      if (size === 'md') {
        return {
          bodyLevel: 4,
          pt: 4,
          pl: direction === 'vertical' ? 32 : 8,
        } as const;
      }

      return {
        bodyLevel: 5,
        pt: direction === 'vertical' ? 2 : 3,
        pl: direction === 'vertical' ? 28 : 8,
      } as const;
    },
  })
);
