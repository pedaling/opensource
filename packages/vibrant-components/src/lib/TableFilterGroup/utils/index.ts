import type { DateFilterOperator, MultiSelectFilterOperator, Option, StringFilterOperator } from '../types';

export const isValueRequiredOperator = (op: string) => op === 'empty' || op === 'notEmpty';

export const isStringFilterValid = (filter: { value: string; operator: StringFilterOperator }) =>
  Boolean(filter.value) || isValueRequiredOperator(filter.operator);

export const isDateFilterValid = (filter: { value: Date[]; operator: DateFilterOperator }) => {
  if (isValueRequiredOperator(filter.operator)) return true;

  if (filter.operator === 'between') return filter.value.length >= 2;

  return filter.value.length >= 1;
};

export const isMultiSelectFilterValid = (filter: { value: Option['value'][]; operator: MultiSelectFilterOperator }) =>
  Boolean(filter.value.length > 0) || isValueRequiredOperator(filter.operator);
