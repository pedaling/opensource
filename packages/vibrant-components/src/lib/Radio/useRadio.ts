import { useEffect, useRef } from 'react';
import { isDefined } from '@vibrant-ui/utils';
import { useRadioGroup } from '../RadioGroupField/context/RadioGroupContext';
import type { RadioProps } from './RadioProps';

export function useRadio({ value, disabled, checked }: Pick<RadioProps, 'checked' | 'disabled' | 'value'>) {
  const { name, value: checkedValue, onValueChange, disabled: groupDisabled, state } = useRadioGroup();

  const isChecked = value === checkedValue;
  const isDisabled = groupDisabled || disabled;
  const prevCheckedRef = useRef<boolean>();

  useEffect(() => {
    if (!isDefined(checked) || checked === prevCheckedRef.current) {
      return;
    }

    if (checked) {
      onValueChange(value);
    }

    if (!checked && isChecked) {
      onValueChange('');
    }
  }, [checked, isChecked, onValueChange, value]);

  useEffect(() => {
    prevCheckedRef.current = checked;
  }, [checked, isChecked]);

  const handleCheckedChange = () => {
    if (isChecked) {
      return;
    }

    onValueChange(isChecked ? '' : value);
  };

  return {
    name,
    isChecked,
    isDisabled,
    onChange: handleCheckedChange,
    state,
  };
}
