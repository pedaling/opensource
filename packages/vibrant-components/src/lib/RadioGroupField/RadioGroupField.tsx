import { useEffect, useState } from 'react';
import { useCallbackRef } from '@vibrant-ui/utils';
import { RadioGroupProvider } from './context/RadioGroupContext';
import type { RadioGroupContextValue } from './context/RadioGroupContext';
import { withRadioGroupFieldVariation } from './RadioGroupFieldProps';

export const RadioGroupField = withRadioGroupFieldVariation(({ defaultValue, onValueChange, children, ...props }) => {
  const [value, setValue] = useState(defaultValue);

  useEffect(() => {
    setValue(defaultValue);
  }, [defaultValue]);

  const handleValueChange = useCallbackRef<RadioGroupContextValue['onValueChange']>(event => {
    setValue(event.value);

    onValueChange?.(event);
  });

  return (
    <RadioGroupProvider value={value} onValueChange={handleValueChange} {...props}>
      {children}
    </RadioGroupProvider>
  );
});
