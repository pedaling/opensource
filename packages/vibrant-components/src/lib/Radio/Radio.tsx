import { Box, PressableBox, isNative } from '@vibrant-ui/core';
import { Icon } from '@vibrant-ui/icons';
import { Body } from '../Body';
import { HStack } from '../HStack';
import { useRadioSize } from '../RadioGroupField/context/RadioSizeContext';
import {
  withRadioDescriptionVariation,
  withRadioIconVariation,
  withRadioLabelVariation,
  withRadioVariation,
} from './RadioProps';
import { useRadio } from './useRadio';

export const Radio = withRadioVariation(
  ({ innerRef, checked, value, label, description, size: sizeProp, direction, flexDirection, disabled }) => {
    const { name, isChecked, isDisabled, onChange } = useRadio({ value, checked, disabled });
    const radioGroupSize = useRadioSize();
    const size = radioGroupSize ?? sizeProp;

    return (
      <PressableBox
        as="label"
        flexDirection={flexDirection}
        width="100%"
        disabled={isDisabled}
        cursor={isDisabled ? 'default' : 'pointer'}
        onClick={isNative ? onChange : undefined}
        tabIndex={isNative ? 0 : -1}
      >
        <HStack as="span">
          <Box as="span">
            {!isNative ? (
              <Box
                ref={innerRef}
                as="input"
                name={name}
                type="radio"
                value={value}
                position="absolute"
                opacity={0}
                checked={isChecked}
                disabled={isDisabled}
                onChange={onChange}
                top={0}
                right={0}
                bottom={0}
                left={0}
              />
            ) : null}
            <RadioIcon size={size} checked={isChecked} disabled={isDisabled} />
          </Box>
          {label ? <RadioLabel label={label} disabled={disabled} size={size} /> : null}
        </HStack>

        {description ? (
          <RadioDescription description={description} disabled={isDisabled} size={size} direction={direction} />
        ) : null}
      </PressableBox>
    );
  }
);

export const RadioLabel = withRadioLabelVariation(({ bodyLevel, pl, pt, label, disabled }) => (
  <Body level={bodyLevel} pl={pl} pt={pt} color={disabled ? 'onView3' : 'onView1'}>
    {label}
  </Body>
));

export const RadioDescription = withRadioDescriptionVariation(({ bodyLevel, pl, pt, description, disabled }) => (
  <Body level={bodyLevel} pl={pl} pt={pt} color={disabled ? 'onView3' : 'onView2'}>
    {description}
  </Body>
));

export const RadioIcon = withRadioIconVariation(({ size, checked, disabled }) =>
  checked ? (
    <Icon.ToggleOn.Fill fill={disabled ? 'onView3' : 'onViewPrimary'} size={size} />
  ) : (
    <Icon.ToggleOff.Thin fill="onView3" size={size} />
  )
);
