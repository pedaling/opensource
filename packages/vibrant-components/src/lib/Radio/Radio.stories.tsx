import { useId } from 'react';
import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { Box } from '@vibrant-ui/core';
import { Body } from '../Body';
import { Pressable } from '../Pressable';
import { RadioGroupField } from '../RadioGroupField/RadioGroupField';
import { VStack } from '../VStack';
import { Radio } from './Radio';
import { useRadio } from './useRadio';

export default {
  title: 'Radio',
  component: Radio,
  args: {
    label: 'Radio',
    description: 'Description',
    value: '1',
  },
} as ComponentMeta<typeof Radio>;

export const Basic: ComponentStory<typeof Radio> = props => (
  <RadioGroupField name="radioGroup">
    <VStack spacing={8} width={300} alignHorizontal="start">
      <Radio {...props} />
      <Radio value="2" label="Second" />
      <Radio value="3" label="Third" />
    </VStack>
  </RadioGroupField>
);

const RadioButton = ({ value, label }: { value: string; label: string }) => {
  const { name, isChecked, isDisabled, onChange } = useRadio({ value });

  const id = useId();

  return (
    <Pressable
      as="span"
      width={100}
      disabled={isDisabled}
      backgroundColor={isChecked ? 'primary' : 'background'}
      p={10}
      rounded="lg"
    >
      <>
        <Box
          as="input"
          id={id}
          value={value}
          name={name}
          type="radio"
          checked={isChecked}
          onChange={onChange}
          opacity={0}
          position="absolute"
          top={0}
          left={0}
          width="100%"
          height="100%"
        />
        <Body as="label" level={1} htmlFor={id}>
          {label}
        </Body>
      </>
    </Pressable>
  );
};

export const Custom: ComponentStory<typeof Radio> = () => (
  <RadioGroupField name="radioGroup">
    <RadioButton value="1" label="First" />
    <RadioButton value="2" label="Second" />
    <RadioButton value="3" label="Third" />
  </RadioGroupField>
);
