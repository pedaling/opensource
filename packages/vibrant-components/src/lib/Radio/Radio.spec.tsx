import { waitFor } from '@testing-library/react';
import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { RadioGroupField } from '../RadioGroupField';
import { Radio } from './Radio';

describe('<Radio />', () => {
  const { render } = createReactRenderer();
  let renderer: ReactRenderer;

  describe('when checked is true', () => {
    beforeEach(() => {
      renderer = render(
        <RadioGroupField name="test">
          <Radio value="radio" checked={true} />
        </RadioGroupField>
      );
    });

    it('radio should be checked', async () => {
      await waitFor(() => expect(renderer.getByRole('radio')).toHaveProperty('checked', true));
    });
  });

  describe('when size is md', () => {
    beforeEach(() => {
      renderer = render(
        <RadioGroupField name="test">
          <Radio value="radio" size="md" />
        </RadioGroupField>
      );
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when size is sm', () => {
    beforeEach(() => {
      renderer = render(
        <RadioGroupField name="test">
          <Radio value="radio" size="sm" />
        </RadioGroupField>
      );
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });
});
