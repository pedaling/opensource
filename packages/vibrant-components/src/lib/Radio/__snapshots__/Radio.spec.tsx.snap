// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> when size is md match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 100%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: absolute;
  opacity: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 24px;
  height: 24px;
  fill: #0000000d;
}

<div>
  <div
    class="emotion-0"
  >
    <label
      class="emotion-1"
      tabindex="-1"
    >
      <span
        class="emotion-2"
      >
        <input
          class="emotion-3"
          name="test"
          type="radio"
          value="radio"
        />
        <svg
          class="emotion-4"
          data-testid="toggleoff-thin"
          style="opacity: 1; transition-property: opacity; transition-duration: 200ms; transition-timing-function: cubic-bezier(0.5, 1, 0.89, 1);"
          viewBox="0 0 24 24"
        >
          <path
            class="emotion-2"
            d="M12 2.75c5.1 0 9.25 4.15 9.25 9.25S17.1 21.25 12 21.25 2.75 17.1 2.75 12 6.9 2.75 12 2.75ZM12 1C5.9 1 1 5.9 1 12s4.9 11 11 11 11-4.9 11-11S18.1 1 12 1Z"
          />
        </svg>
      </span>
    </label>
  </div>
</div>
`;

exports[`<Radio /> when size is sm match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 100%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: absolute;
  opacity: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 20px;
  height: 20px;
  fill: #0000000d;
}

<div>
  <div
    class="emotion-0"
  >
    <label
      class="emotion-1"
      tabindex="-1"
    >
      <span
        class="emotion-2"
      >
        <input
          class="emotion-3"
          name="test"
          type="radio"
          value="radio"
        />
        <svg
          class="emotion-4"
          data-testid="toggleoff-thin"
          style="opacity: 1; transition-property: opacity; transition-duration: 200ms; transition-timing-function: cubic-bezier(0.5, 1, 0.89, 1);"
          viewBox="0 0 24 24"
        >
          <path
            class="emotion-2"
            d="M12 2.75c5.1 0 9.25 4.15 9.25 9.25S17.1 21.25 12 21.25 2.75 17.1 2.75 12 6.9 2.75 12 2.75ZM12 1C5.9 1 1 5.9 1 12s4.9 11 11 11 11-4.9 11-11S18.1 1 12 1Z"
          />
        </svg>
      </span>
    </label>
  </div>
</div>
`;
